### PR TITLE
Endtoend: stress tests for VTGate FOREIGN KEY support

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Adding `endtoend` stress tests for VTGate `FOREIGN KEY` functionality.

~~At this time VTGate does not implement complete support for al reference actions. This test will therefore fail in the interim.~~

Excerpt from test comment:

This endtoend test is designd to validate VTGate's FOREIGN KEY implementation for unsharded/single-sharded/shard-scope, meaning we expect foreign key constraints to be limited to a shard (related rows can never be on diffrent shards).

This test validates `NO ACTION`, `CASCADE` and `SET NULL` reference actions.
VTGate's support for foreign keys includes:
- Analyzing the foreign key constraints in a keyspace.
- Rejecting `INSERT` statements for child table when there's no matching row on a parent table.
- Handling `DELETE` and `UPDATE` statements on a parent table according to the reference action on all children.
  Specifically, this means for example that VTGate will handle a ON DELETE CASCADE in Vitess plane. It will first delete rows
  from the child (recursive operation) before deleting the row on the parent. As result, the underlying MySQL server will have
  nothing to cascade.

The design of this test is as follows:
- Create a cluster with PRIMARY and REPLICA tablets
- Given this structure of tables with foreign key constraints:
```
  stress_parent
  +- stress_child
     +- stress_grandchild
  +- stress_child2
```
- Create these tables. Then, on the MySQL replica, remove the foreign key constraints.
- Static test:
  - Randomly populate all tables via highly-contentive `INSERT/UPDATE/DELETE` statements
  - Validate collected metrics match actual table data
  - Validate foreign key constraints integrity
- Workload test:
  - Initially populate tables as above
  - Run a high contention workload where multiple connections issue random `INSERT/UPDATE/DELETE` on all related tables
  - Validate collected metrics match actual table data
  - Validate foreign key constraints integrity on MySQL primary
  - Validate foreign key constraints integrity on MySQL replica
  - Compare data on primary & replica

We of course know that foreign key integrity is maintained on the MySQL primary. However, the replica does not have the matching constraints. Since cascaded (SET NULL, CASCADE) writes are handled internally by InnoDB and not written to the binary log, any cascaded writes on the primary are lost, and the replica is unaware of those writes. Without VTGate intervention, we expect the replica to quickly diverge from the primary, and in fact in all likelyhood replication will break very quickly.
However, if VTGate implements the cascading rules correctly, the primary MySQL server will never have any actual cascades, and so cascaded writes are all accounted for in the binary logs, which means we can expect the replica to be compliant with the primary.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/11975

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/test/endtoend/vtgate/foreignkey/stress/fk_stress_test.go [hunk 1]
```diff
@@ -0,0 +1,882 @@
...
+			hint_col varchar(64) not null default '',
+			created_timestamp timestamp not null default current_timestamp,
+			updates int unsigned not null default 0,
+			PRIMARY KEY (id),
+			key parent_id_idx(parent_id),
+			key created_idx(created_timestamp),
+			key updates_idx(updates),
+			CONSTRAINT child_parent_fk FOREIGN KEY (parent_id) REFERENCES stress_parent (id) ON DELETE %s
+		) ENGINE=InnoDB
+		`,
```
Comment (harshit-gangal) [inline go/test/endtoend/vtgate/foreignkey/stress/fk_stress_test.go:178]: on update will be `RESTRICT` is that intentional?

Comment (shlomi-noach) [reply go/test/endtoend/vtgate/foreignkey/stress/fk_stress_test.go:178]: You're right. The entire test is only testing `ON DELETE`. I'll add `ON UPDATE`.

Comment (shlomi-noach) [reply go/test/endtoend/vtgate/foreignkey/stress/fk_stress_test.go]: The tests now validate all combinations of `ON DELETE` cross `ON UPDATE`. The tests validate `ON UPDATE CASCADE` and `ON UPDATE SET NULL` by issuing queries that modify the foreign key parent column (`id` in our case).

File: go/test/endtoend/vtgate/foreignkey/stress/fk_stress_test.go [hunk 2]
```diff
@@ -0,0 +1,1180 @@
...
+		}
+		select {
+		case <-ctx.Done():
+			assert.FailNow(t, "timeout waiting for replica to catch up")
+			return
+		case <-time.After(time.Second):
+			//
+		}
+	}
+}
```
Comment (GuptaManan100) [inline go/test/endtoend/vtgate/foreignkey/stress/fk_stress_test.go:430]: There is a function in `cluster` package which does something very similar. The function can be called like so - `cluster.WaitForReplicationPos(t, primaryTab, replicaTab, "localhost", 60.0)`
That function gets the position for the primary and waits for upto the amount of seconds provided for the replica tablet to be atleast as advanced as that position.

**Review by GuptaManan100 — APPROVED** [review]
Rest looks good to me!

File: go/test/endtoend/vtgate/foreignkey/stress/fk_stress_test.go (hunk 2 above)
Comment (shlomi-noach) [reply go/test/endtoend/vtgate/foreignkey/stress/fk_stress_test.go]: OK, converged with `cluster.WaitForReplicationPos`. But to do that, I also made changes to `cluster.WaitForReplicationPos`:

- The function `waitStep()` was incorrect as well as _used_ incorrectly. Removed it and used idiomatic `context.WithCancel` in `WaitForReplicationPos`
- `hostname` is assumed to be `"localhost"` as this is hard coded in the cluster creation anyway.
- Added `validateReplication` function argument. When replication is broken, there is generally no point in waiting for the position. This is opt-in (by default, check is not made).

Comment (GuptaManan100) [reply go/test/endtoend/vtgate/foreignkey/stress/fk_stress_test.go]: Sweet! Thank-you!

**Comment by shlomi-noach** [comment]
@harshit-gangal another view please?

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**